### DISCUSSION TITLE
importing a third-party module in my script

### DISCUSSION THREAD

**Discussion by gagagagawa** [micropython/micropython#12328] [Libraries & Drivers] [answered]
Hi, im new to micropython but ill give my best to describe my issue

i have setup my Arduino Nano 33 BLE with the micropython firmware and im using the pyboard.py tool to execute some code and test my setup so far. I can run commands so the communication with the device seems to be setup correctly.

Now i want to use this module https://github.com/robert-hh/ads1x15 to use my ads1115.
I used: python3 pyboard.py -f cp /path/to/ads1x15.py :third-party/ads1x15.py to copy my module that i downloaded to the arduino.
If i use pytyhon3 pyboard.py -f ls third-party/ it shows the ads1x15.py script so this also seemed to work.

But when i now use: python3 pyboard.py -c "import ads1x15.py" it can not find the module.
What i did not understand yet is if he really searches for the module when importing in the /third-party directory because 
import sys
print(sys.path)
only returns ['','.frozen'] and i dont really know how to directories are setup in the firmware

sorry if its just some simple setup for micropython im forgetting so he also searches in the directory but i dont really know how to proceed from here and was hoping for some help

Thank you in advance and i hope i described my problem good enough

**Comment by robert-hh** [comment]
Better use mpremote.py. It's either here https://github.com/micropython/micropython/tree/master/tools/mpremote
or can be installed with `pip install mpremote`.
In order to run code, it has to be in the path list of sys.path. This list can be extended e.g. with `sys.path.append(<new path>)`.

**Reply by gagagagawa** [reply]
Im now using mpremote, thank you.
Since i already manually installed it with pyboard.py -f cp  i could sys.path.append('third-party/') and the import my module
So this works and with my manual installation process this would work with other modules aswell. Thanks!
And also thanks for the ADS1x15 module, it will of course save me alot of work when i have my script setup for the Nano 33 BLE 😃

**Comment by mattytrentini** [comment]
Install it using `mip` (via `mpremote` since your device doesn't have wifi support). See [Package Management](https://docs.micropython.org/en/latest/reference/packages.html) for details.

(The first line below is key!)

```bash
❯ mpremote mip install github:robert-hh/ads1x15/ads1x15.py
Install github:robert-hh/ads1x15/ads1x15.py
Downloading github:robert-hh/ads1x15/ads1x15.py to /lib
Installing: /lib/ads1x15.py
Done

❯ mpremote
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] to exit this shell
>>> import ads1x15
>>> help(ads1x15)
object <module 'ads1x15' from '/lib/ads1x15.py'> is of type module
  ADS1015 -- <class 'ADS1015'>
  ADS1114 -- <class 'ADS1114'>
  ADS1115 -- <class 'ADS1115'>
  _CHANNELS -- {(2, None): 24576, (3, None): 28672, (0, 1): 0, (0, 3): 4096, (1, 3): 8192, (2, 3): 12288, (0, None): 16384, (1, None): 20480}
  __file__ -- /lib/ads1x15.py
  __name__ -- ads1x15
  _RATES -- (0, 32, 64, 96, 128, 160, 192, 224)
  _GAINS -- (0, 512, 1024, 1536, 2048, 2560)
  ADS1113 -- <class 'ADS1113'>
  _GAINS_V -- (6.144, 4.096, 2.048, 1.024, 0.512, 0.256)
  time -- <module 'time'>
>>>
```

**Reply by gagagagawa** [reply]
so i already tried using mpremote mip install before, in fact exactly the command you executed
$ mpremote mip install github:robert-hh/ads1x15/ads1x15.py
and then i get this:
Install github:robert-hh/ads1x15/ads1x15.py
mpremote: Unable to find lib dir in sys.path, use --target to override
i then added a lib directory by using
$ mpremote fs mkdir lib (which is also listed by "mpremote fs ls" then)
and then executed
$ mpremote mip install --target=lib github:robert-hh/ads1x15/ads1x15.py
this was succesful. I now know how to install using mip, this ofc helps alot, so thanks.

I now have the following problem, i dont know if it is intended that neither lib or third-party is in sys.path so i would need to add them at the start of my script before importing from these directories. Is it possible to change the sys.path "permanently", so when resetting the connection and executing a script or opening new repl connection, it will already be added to sys.path?

Sorry, i have the feeling this is more of a python question and missing knowledge on my side about python in general but it would be convenient to set it up like described.

**Reply by glenn20** [reply, marked as answer]
On esp32, esp8266 and rp2; `/lib` is added to sys.path during firmware initialisation (in `main.c`). It seems that is not universally true for all ports, so you could add it in your `main.py`: ```import sys; sys.path.append("/lib")```.

**Reply by gagagagawa** [reply]
Sorry for the late reply. Yeah I also think this will be my way to go if I keep using this port but i have the feeling it would be better to switch to my esp8266 now. Thank you for the help!
